### PR TITLE
Allow overriding accepted status codes for cached URIs

### LINKED ISSUE #840: lychee ignores http accepted status codes when using cache
It looks like lychee doesn't work with cache and accepted status codes properly.

See the example:

Prepare the data + directory:

```bash
mkdir /tmp/test
cd /tmp/test

cat > lychee.toml << EOF
accept = [ 999 ]
EOF

cat > README.md << EOF
[https://www.linkedin.com/in/petrruzicka/](https://www.linkedin.com/in/petrruzicka/)
EOF
```

Run lychee (first run is working fine):

```bash
$ lychee --cache .
⠉ 1/2 [00:00:00] ████████████░░░░░░░░░░░░░ ✔ [999] https://www.linkedin.com/in/petrruzicka/ | OK (999 <unknown status co
⠉ 2/2 [00:00:00] █████████████████████████ ✔ [999] https://www.linkedin.com/in/petrruzicka/ | OK (999 <unknown status co
🔍 2 Total ✅ 2 OK 🚫 0 Errors
```

Second run is failing - I still have `accept = [ 999 ]` in `lychee.toml` :-(

```bash
$ lychee --cache .
Issues found in 1 input. Find details below.

[./README.md]:
↻ [999] https://www.linkedin.com/in/petrruzicka/ | Cached: Error (cached)

🔍 2 Total ✅ 0 OK 🚫 2 Errors (HTTP:2)
```

Cache details:

```bash
❯ cat .lycheecache
https://www.linkedin.com/in/petrruzicka/,999,1669292835
```

Seems to be a bug...

Thank you for looking at it.

### PR DESCRIPTION
**Pull request by mre**
Fixes #840

### CODE REVIEW

File: lychee-lib/src/types/status.rs [hunk 1]
```diff
@@ -84,6 +84,36 @@ impl Status {
...
+    /// because they are provided by the user and can be invalid according to
+    /// the HTTP spec and IANA, but the user might still want to accept them.
+    #[must_use]
+    pub fn from_cache_status(s: CacheStatus, accepted: Option<HashSet<u16>>) -> Self {
+        match s {
+            CacheStatus::Ok(code) => {
+                if accepted.map(|a| a.contains(&code)) == Some(true) {
+                    return Self::Cached(CacheStatus::Ok(code));
+                };
+                Self::Cached(CacheStatus::Error(Some(code)))
```
Comment (mre) [inline lychee-lib/src/types/status.rs:103]: Not sure if we should mark cached status codes, which are not in the set of accepted status codes as errors. Any thoughts? 🤔

Comment (mre) [reply lychee-lib/src/types/status.rs:106]: Added a comment about that. I think it's an easily revertible change if we find that it's not what we want.

**Comment by MichaIng** [comment]
This throws now an error for all cached results, even that those did not throw an error before: https://github.com/MichaIng/DietPi-Docs/actions/runs/4247085616/jobs/7384705737

File: lychee-bin/tests/cli.rs [hunk 2]
```diff
@@ -771,6 +771,81 @@ mod cli {
...
+        assert!(data.contains(&format!("{}/,200", mock_server_ok.uri())));
+        assert!(data.contains(&format!("{}/,418", mock_server_teapot.uri())));
+        assert!(data.contains(&format!("{}/,500", mock_server_server_error.uri())));
+
+        // run again to verify cache behavior
+        // this time accept 418 and 500 as valid status codes
+        test_cmd
+            .arg("--no-progress")
+            .arg("--accept")
+            .arg("200,418,500")
```
Comment (MichaIng) [inline lychee-bin/tests/cli.rs:831]: I think it does not require `200` to be accepted explicitly. I suggest to remove this from the test and assure that code `200` never needs to be explicitly accepted, to match the behaviour of uncached checks.

**Comment by mre** [comment]
Ouch. Thanks for the heads-up. That's a bit of a shame. I've created a hotfix for that. Can you give that one a shot before I release it? 🙇

File: lychee-bin/tests/cli.rs (hunk 2 above)
Comment (mre) [reply lychee-bin/tests/cli.rs:831]: Thanks, you're right. Updated in #958.

**Pull request merged**